source/common/network/dns_impl.cc

```diff
@@ -1,97 +1,126 @@
 #include "dns_impl.h"
 
 #include "common/common/assert.h"
-#include "common/event/libevent.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
 
+#include "ares.h"
 #include "event2/event.h"
 
 namespace Network {
 
-DnsResolverImpl::DnsResolverImpl(Event::DispatcherImpl& dispatcher) : dispatcher_(dispatcher) {
-  // This sets us up to receive signals on an fd when async DNS resolved are completed.
-  sigset_t mask;
-  sigemptyset(&mask);
-  sigaddset(&mask, Event::Libevent::Global::DNS_SIGNAL_ID);
-  signal_fd_ = signalfd(-1, &mask, SFD_NONBLOCK);
-  RELEASE_ASSERT(-1 != signal_fd_);
-
-  event_assign(&signal_read_event_, &dispatcher_.base(), signal_fd_,
-               EV_READ | EV_PERSIST, [](evutil_socket_t, short, void* arg) -> void {
-                 static_cast<DnsResolverImpl*>(arg)->onSignal();
-               }, this);
-
-  event_add(&signal_read_event_, nullptr);
+DnsResolverImpl::DnsResolverImpl(Event::DispatcherImpl& dispatcher)
+    : dispatcher_(dispatcher),
+      timer_(dispatcher.createTimer([this] { onEventCallback(ARES_SOCKET_BAD, 0); })) {
+  // This is also done in main(), to satisfy the requirement that c-ares is
+  // initialized prior to threading. The additional call to ares_library_init()
+  // here is a nop in normal execution, but exists for testing where we don't
+  // launch via main().
+  ares_library_init(ARES_LIB_INIT_ALL);
+  ares_options options;
+  initializeChannel(&options, 0);
 }
 
 DnsResolverImpl::~DnsResolverImpl() {
-  close(signal_fd_);
-  event_del(&signal_read_event_);
+  timer_->disableTimer();
+  ares_destroy(channel_);
+  ares_library_cleanup();
 }
 
-void DnsResolverImpl::onSignal() {
-  while (true) {
-    signalfd_siginfo signal_info;
-    ssize_t rc = read(signal_fd_, &signal_info, sizeof(signal_info));
-    if (rc == -1 && errno == EAGAIN) {
-      break;
-    }
-
-    RELEASE_ASSERT(rc == sizeof(signal_info));
-    PendingResolution* pending_resolution =
-        reinterpret_cast<PendingResolution*>(signal_info.ssi_ptr);
+void DnsResolverImpl::initializeChannel(ares_options* options, int optmask) {
+  options->sock_state_cb = [](void* arg, int fd, int read, int write) {
+    static_cast<DnsResolverImpl*>(arg)->onAresSocketStateChange(fd, read, write);
+  };
+  options->sock_state_cb_data = this;
+  ares_init_options(&channel_, options, optmask | ARES_OPT_SOCK_STATE_CB);
+}
 
-    std::list<Address::InstancePtr> address_list;
-    addrinfo* result = pending_resolution->async_cb_data_.ar_result;
-    while (result != nullptr) {
+void DnsResolverImpl::PendingResolution::onAresHostCallback(int status, hostent* hostent) {
+  if (status == ARES_EDESTRUCTION) {
```

Comment:
This owned business is a little confusing. I'm guessing ARES_EDESTRUCTION happens when we destroy the channel with a pending request? If so, can we just put a comment to that effect?